### PR TITLE
Translucent panel

### LINKED ISSUE #182: IES_Tookit: Push/Pull Translucent shade elements
#### Description:

IES can now model translucent shading elements and initial tests suggest that these can be imported via GEM (_although the validity of the solar shading calculation of the imported geometry is TBC_) so we should develop functionality to be able to Push/Pull BHoM objects to/from GEM as translucent shade objects.

GEM format defines a translucent shade using:
```
Type = 1
Subtype = 2102
```

Relates to BHoM.Environment_oM issue: [#1334](https://github.com/BHoM/BHoM/issues/1334)

### PR DESCRIPTION
**Pull request by robinflyman**
Closes #182

### CODE REVIEW

File: IES_Engine/Convert/Environment/Panel.cs [hunk 1]
```diff
@@ -53,14 +53,25 @@ public static List<string> ToIESShading(this List<Panel> panelsAsShade, Settings
 
             for (int x = 0; x < panels.Count; x++)
             {
+                bool isTranslucent = panels[x].Type != PanelType.Shade;
```
Comment (FraserGreenroyd) [inline IES_Engine/Convert/Environment/Panel.cs:56]: Put this before the for loop and set it as `panels.First().Type` instead of `panels[x].Type`

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

File: IES_Adapter/CRUD/Read.cs [hunk 2]
```diff
@@ -120,8 +125,8 @@ private List<Panel> ReadPanels()
                 for (int x = 0; x < nextIndex; x++)
                     space.Add(iesStrings[x]);
 
-                if (panelType == PanelType.Shade)
-                    panels.Add(space.FromIESShading(_settingsIES)); //Make a shade panel
+                if (panelType == PanelType.Shade || panelType == PanelType.TranslucentShade)
+                    panels.Add(space.FromIESShading(_settingsIES,panelType)); //Make a shade panel
```
Comment (FraserGreenroyd) [inline IES_Adapter/CRUD/Read.cs:129]:
```suggestion
                    panels.Add(space.FromIESShading(_settingsIES, panelType)); //Make a shade panel
```

File: IES_Engine/Convert/Environment/Panel.cs [hunk 3]
```diff
@@ -98,7 +109,7 @@ public static List<string> ToIESShading(this List<Panel> panelsAsShade, Settings
         [Input("iesSpace", "The IES representation of a space")]
         [Input("settingsIES", "The IES settings to use with the IES adapter")]
         [Output("panelsAsSpace", "BHoM Environment Space")]
-        public static Panel FromIESShading(this List<string> iesPanel, SettingsIES settingsIES)
+        public static Panel FromIESShading(this List<string> iesPanel, SettingsIES settingsIES, PanelType panelType)
```
Comment (FraserGreenroyd) [inline IES_Engine/Convert/Environment/Panel.cs:112]: Update `line 137` to utilise this `PanelType` input

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Review by RMildiner — APPROVED** [review]
Tested the GEM file creation using a GH script and then imported the GEM into IES. The translucent shade component appeared in the correct place and was recognised as translucent by IES straight away. Seems to work.

**Review by jamesramsden-bh — CHANGES_REQUESTED** [review]
Pushing to IES: Pass. Confirmed with @RMildiner that a pushed model opens in IES correctly.

Pulling to IES: Fail. The 'opening' within the shade is offset by a vector of approximately 8m,8m,8m. This issue is not evident when using a non-translucent shade. (Test GEM files for translucent and non-translucent within the zip file attached.) 
[translucent push and pull tests.zip](https://github.com/BHoM/IES_Toolkit/files/8016726/translucent.push.and.pull.tests.zip)

**Comment by jamesramsden-bh** [comment]
Fail from previous comment now fixed. @FraserGreenroyd now ready for review. 
@robinflyman no action needed from you right now.

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
The use of the newer C# tuples, though very nice, isn't quite in line with the rest of the BHoM. As such, it'd be better to stick with the `Output<t1, tn>` solutions previously implemented.

However, I think we might want to update `FilterPanelsByType` method in the Environment_Engine, and allow it to take in a multiple params of types to filter but also group. I'll raise an issue for that to be tackled as part of this work but before this can be merged.

Then we can rectify this PR to be more in line with what's expected 😄

File: IES_Engine/Convert/Environment/Panel.cs [hunk 4]
```diff
@@ -149,11 +160,16 @@ public static Panel FromIESShading(this List<string> iesPanel, SettingsIES setti
 
             if (settingsIES.PullOpenings)
             {
-                //Fix the openings now
-                for (int x = 0; x < panel.Openings.Count; x++)
+                //This if-statement is a slightly hacky fix for the IES implementation of translucent shades,
```
Comment (FraserGreenroyd) [inline IES_Engine/Convert/Environment/Panel.cs:163]:
```suggestion
                //This if-statement is a fix for the IES implementation of translucent shades, 
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

File: IES_Engine/Convert/Environment/Panel.cs (hunk 3 above)
Comment (FraserGreenroyd) [inline IES_Engine/Convert/Environment/Panel.cs:112]:
```suggestion
        [PreviousVersion("5.1", "BH.Engine.Adapters.IES.Convert.FromIESShading(System.Collections.Generic.List<System.String>, BH.oM.IES.Settings.SettingsIES)")]
        public static Panel FromIESShading(this List<string> iesPanel, SettingsIES settingsIES, PanelType panelType)
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Review by jamesramsden-bh — APPROVED** [review]
Tested against translucent shading sections of testing procedure, LGTM.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check installer

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 5199266217

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 5199267343

**Review by FraserGreenroyd — APPROVED** [review]
Tested through on test procedure by others, code reviewed by me, updates required to engine have been implemented and are working happily. Versioning and installer checks are happy, so deploying to alpha for further testing.

**Pull request merged**